<center>
    <img src="https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>

# Working with a real world data-set using SQL and Python

Estaimted time needed: **30** minutes

## Objectives

After complting this lab you will be able to:

*   Understand the dataset for Chicago Public School level performance
*   Store the dataset in SQLite database.
*   Retrieve metadata about tables and columns and query data from mixed case columns
*   Solve example problems to practice your SQL skills including using built-in database functions


## Chicago Public Schools - Progress Report Cards (2011-2012)

The city of Chicago released a dataset showing all school level performance data used to create School Report Cards for the 2011-2012 school year. The dataset is available from the Chicago Data Portal: [https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

This dataset includes a large number of metrics. Start by familiarizing yourself with the types of metrics in the database: [https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?download=true](https://data.cityofchicago.org/api/assets/AAD41A13-BE8A-4E67-B1F5-86E711E09D5F?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&download=true&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

**NOTE**:

Do not download the dataset directly from City of Chicago portal. Instead download a static copy which is a more database friendly version from this <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">link</a>.

Now review some of its contents.


### Connect to the database

Let us now load the ipython-sql  extension and establish a connection with the database

##### The syntax for connecting to magic sql using sqllite is

**%sql sqlite://DatabaseName**

where DatabaseName will be your **.db** file


In [1]:
import csv, sqlite3

con = sqlite3.connect("RealWorldData.db")
cur = con.cursor()

In [2]:

!pip install -q pandas==1.1.5

In [3]:
%load_ext sql

In [4]:
%sql sqlite:///RealWorldData.db

'Connected: @RealWorldData.db'

### Store the dataset in a Table

##### In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. To analyze the data

##### using SQL, it first needs to be stored in the database.

##### We will first read the csv files  from the given url  into pandas dataframes

##### Next we will be using the  df.to_sql() function to convert each csv file  to a table in sqlite  with the csv data loaded in it.


In [7]:
import pandas
df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv")
df.to_sql("CENSUS_DATA", con, if_exists='replace', index=False,method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv")
df.to_sql("CHICAGO_CRIME_DATA", con, if_exists='replace', index=False, method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv")
df.to_sql("CHICAGO_PUBLIC_SCHOOLS_DATA", con, if_exists='replace', index=False, method="multi")




Double-click **here** for the solution.

<!--

import pandas
df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv")
df.to_sql("CENSUS_DATA", con, if_exists='replace', index=False,method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv")
df.to_sql("CHICAGO_CRIME_DATA", con, if_exists='replace', index=False, method="multi")

df = pandas.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv")
df.to_sql("CHICAGO_PUBLIC_SCHOOLS_DATA", con, if_exists='replace', index=False, method="multi")

-->


### Query the database system catalog to retrieve table metadata

##### You can verify that the table creation was successful by retrieving the list of all tables in your schema and checking whether the SCHOOLS table was created


In [8]:
# type in your query to retrieve list of all tables in the database
%sql SELECT name FROM sqlite_master WHERE type='table'

 * sqlite:///RealWorldData.db
Done.


[('CENSUS_DATA',), ('CHICAGO_CRIME_DATA',), ('CHICAGO_PUBLIC_SCHOOLS_DATA',)]

Double-click **here** for a hint

<!--
In sqlite the system catalog table called sqlite_master contains the table metadata
-->


Double-click **here** for the solution.

<!-- Solution using SQLite:

%sql SELECT name FROM sqlite_master WHERE type='table'

-->


### Query the database system catalog to retrieve column metadata

##### The SCHOOLS table contains a large number of columns. How many columns does this table have?


In [48]:
# type in your query to retrieve the number of columns in the SCHOOLS table
%sql SELECT count(name) FROM PRAGMA_TABLE_INFO('CENSUS_DATA');

 * sqlite:///RealWorldData.db
Done.


[(9,)]

Double-click **here** for the solution.

<!-- Solution SQLite:
%sql SELECT count(name) FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->


Now retrieve the the list of columns in SCHOOLS table and their column type (datatype) and length.


In [49]:
# type in your query to retrieve all column names in the SCHOOLS table along with their datatypes and length
%sql SELECT * FROM CENSUS_DATA;

 * sqlite:///RealWorldData.db
Done.


[(1.0, 'Rogers Park', 7.7, 23.6, 8.7, 18.2, 27.5, 23939, 39.0),
 (2.0, 'West Ridge', 7.8, 17.2, 8.8, 20.8, 38.5, 23040, 46.0),
 (3.0, 'Uptown', 3.8, 24.0, 8.9, 11.8, 22.2, 35787, 20.0),
 (4.0, 'Lincoln Square', 3.4, 10.9, 8.2, 13.4, 25.5, 37524, 17.0),
 (5.0, 'North Center', 0.3, 7.5, 5.2, 4.5, 26.2, 57123, 6.0),
 (6.0, 'Lake View', 1.1, 11.4, 4.7, 2.6, 17.0, 60058, 5.0),
 (7.0, 'Lincoln Park', 0.8, 12.3, 5.1, 3.6, 21.5, 71551, 2.0),
 (8.0, 'Near North Side', 1.9, 12.9, 7.0, 2.5, 22.6, 88669, 1.0),
 (9.0, 'Edison Park', 1.1, 3.3, 6.5, 7.4, 35.3, 40959, 8.0),
 (10.0, 'Norwood Park', 2.0, 5.4, 9.0, 11.5, 39.5, 32875, 21.0),
 (11.0, 'Jefferson Park', 2.7, 8.6, 12.4, 13.4, 35.5, 27751, 25.0),
 (12.0, 'Forest Glen', 1.1, 7.5, 6.8, 4.9, 40.5, 44164, 11.0),
 (13.0, 'North Park', 3.9, 13.2, 9.9, 14.4, 39.0, 26576, 33.0),
 (14.0, 'Albany Park', 11.3, 19.2, 10.0, 32.9, 32.0, 21323, 53.0),
 (15.0, 'Portage Park', 4.1, 11.6, 12.6, 19.3, 34.0, 24336, 35.0),
 (16.0, 'Irving Park', 6.3, 13.1, 10.0, 22.4, 31.6, 27249, 34.0),
 (17.0, 'Dunning', 5.2, 10.6, 10.0, 16.2, 33.6, 26282, 28.0),
 (18.0, 'Montclaire', 8.1, 15.3, 13.8, 23.5, 38.6, 22014, 50.0),
 (19.0, 'Belmont Cragin', 10.8, 18.7, 14.6, 37.3, 37.3, 15461, 70.0),
 (20.0, 'Hermosa', 6.9, 20.5, 13.1, 41.6, 36.4, 15089, 71.0),
 (21.0, 'Avondale', 6.0, 15.3, 9.2, 24.7, 31.0, 20039, 42.0),
 (22.0, 'Logan Square', 3.2, 16.8, 8.2, 14.8, 26.2, 31908, 23.0),
 (23.0, 'Humboldt park', 14.8, 33.9, 17.3, 35.4, 38.0, 13781, 85.0),
 (24.0, 'West Town', 2.3, 14.7, 6.6, 12.9, 21.7, 43198, 10.0),
 (25.0, 'Austin', 6.3, 28.6, 22.6, 24.4, 37.9, 15957, 73.0),
 (26.0, 'West Garfield Park', 9.4, 41.7, 25.8, 24.5, 43.6, 10934, 92.0),
 (27.0, 'East Garfield Park', 8.2, 42.4, 19.6, 21.3, 43.2, 12961, 83.0),
 (28.0, 'Near West Side', 3.8, 20.6, 10.7, 9.6, 22.2, 44689, 15.0),
 (29.0, 'North Lawndale', 7.4, 43.1, 21.2, 27.6, 42.7, 12034, 87.0),
 (30.0, 'South Lawndale', 15.2, 30.7, 15.8, 54.8, 33.8, 10402, 96.0),
 (31.0, 'Lower West Side', 9.6, 25.8, 15.8, 40.7, 32.6, 16444, 76.0),
 (32.0, 'Loop', 1.5, 14.7, 5.7, 3.1, 13.5, 65526, 3.0),
 (33.0, 'Near South Side', 1.3, 13.8, 4.9, 7.4, 21.8, 59077, 7.0),
 (34.0, 'Armour Square', 5.7, 40.1, 16.7, 34.5, 38.3, 16148, 82.0),
 (35.0, 'Douglas', 1.8, 29.6, 18.2, 14.3, 30.7, 23791, 47.0),
 (36.0, 'Oakland', 1.3, 39.7, 28.7, 18.4, 40.4, 19252, 78.0),
 (37.0, 'Fuller Park', 3.2, 51.2, 33.9, 26.6, 44.9, 10432, 97.0),
 (38.0, 'Grand Boulevard', 3.3, 29.3, 24.3, 15.9, 39.5, 23472, 57.0),
 (39.0, 'Kenwood', 2.4, 21.7, 15.7, 11.3, 35.4, 35911, 26.0),
 (40.0, 'Washington Park', 5.6, 42.1, 28.6, 25.4, 42.8, 13785, 88.0),
 (41.0, 'Hyde Park', 1.5, 18.4, 8.4, 4.3, 26.2, 39056, 14.0),
 (42.0, 'Woodlawn', 2.9, 30.7, 23.4, 16.5, 36.1, 18672, 58.0),
 (43.0, 'South Shore', 2.8, 31.1, 20.0, 14.0, 35.7, 19398, 55.0),
 (44.0, 'Chatham', 3.3, 27.8, 24.0, 14.5, 40.3, 18881, 60.0),
 (45.0, 'Avalon Park', 1.4, 17.2, 21.1, 10.6, 39.3, 24454, 41.0),
 (46.0, 'South Chicago', 4.7, 29.8, 19.7, 26.6, 41.1, 16579, 75.0),
 (47.0, 'Burnside', 6.8, 33.0, 18.6, 19.3, 42.7, 12515, 79.0),
 (48.0, 'Calumet Heights', 2.1, 11.5, 20.0, 11.0, 44.0, 28887, 38.0),
 (49.0, 'Roseland', 2.5, 19.8, 20.3, 16.9, 41.2, 17949, 52.0),
 (50.0, 'Pullman', 1.5, 21.6, 22.8, 13.1, 38.6, 20588, 51.0),
 (51.0, 'South Deering', 4.0, 29.2, 16.3, 21.0, 39.5, 14685, 65.0),
 (52.0, 'East Side', 6.8, 19.2, 12.1, 31.9, 42.8, 17104, 64.0),
 (53.0, 'West Pullman', 3.3, 25.9, 19.4, 20.5, 42.1, 16563, 62.0),
 (54.0, 'Riverdale', 5.8, 56.5, 34.6, 27.5, 51.5, 8201, 98.0),
 (55.0, 'Hegewisch', 3.3, 17.1, 9.6, 19.2, 42.9, 22677, 44.0),
 (56.0, 'Garfield Ridge', 2.6, 8.8, 11.3, 19.3, 38.1, 26353, 32.0),
 (57.0, 'Archer Heights', 8.5, 14.1, 16.5, 35.9, 39.2, 16134, 67.0),
 (58.0, 'Brighton Park', 14.4, 23.6, 13.9, 45.1, 39.3, 13089, 84.0),
 (59.0, 'McKinley Park', 7.2, 18.7, 13.4, 32.9, 35.6, 16954, 61.0),
 (60.0, 'Bridgeport', 4.5, 18.9, 13.7, 22.2, 31.3, 22694, 43.0),
 (61.0, 'New City', 11.9, 29.0, 23.0, 41.5, 38.9, 12765, 91.0),
 (62.0, 'West El

Double-click **here** for the solution.

<!-- Solution 

%sql SELECT name,type,length(type) FROM PRAGMA_TABLE_INFO('CHICAGO_PUBLIC_SCHOOLS_DATA');

-->


### Questions

1.  Is the column name for the "SCHOOL ID" attribute in upper or mixed case?
2.  What is the name of "Community Area Name" column in your table? Does it have spaces?
3.  Are there any columns in whose names the spaces and paranthesis (round brackets) have been replaced by the underscore character "\_"?


## Problems

### Problem 1

##### How many Elementary Schools are in the dataset?


In [16]:
%sql select count(*) FROM CHICAGO_PUBLIC_SCHOOLS_DATA where "Elementary, Middle, or High School"='ES';

 * sqlite:///RealWorldData.db
Done.


[(462,)]

Double-click **here** for a hint

<!--
Which column specifies the school type e.g. 'ES', 'MS', 'HS'? ("Elementary School, Middle School, High School")
-->


Double-click **here** for another hint

<!--
Does the column name have mixed case, spaces or other special characters?
If so, ensure you use double quotes around the "Name of the Column"
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select count(*) from CHICAGO_PUBLIC_SCHOOLS_DATA where "Elementary, Middle, or High School"='ES'

Correct answer: 462

-->


### Problem 2

##### What is the highest Safety Score?


In [17]:
%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from CHICAGO_PUBLIC_SCHOOLS_DATA;



 * sqlite:///RealWorldData.db
Done.


[(99.0,)]

Double-click **here** for a hint

<!--
Use the MAX() function
-->


Double-click **here** for the solution.

<!-- Hint:

%sql select MAX(Safety_Score) AS MAX_SAFETY_SCORE from CHICAGO_PUBLIC_SCHOOLS_DATA

Correct answer: 99
-->


### Problem 3

##### Which schools have highest Safety Score?


In [19]:
%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where Safety_Score=99.0;



 * sqlite:///RealWorldData.db
Done.


[('Abraham Lincoln Elementary School', 99.0),
 ('Alexander Graham Bell Elementary School', 99.0),
 ('Annie Keller Elementary Gifted Magnet School', 99.0),
 ('Augustus H Burley Elementary School', 99.0),
 ('Edgar Allan Poe Elementary Classical School', 99.0),
 ('Edgebrook Elementary School', 99.0),
 ('Ellen Mitchell Elementary School', 99.0),
 ('James E McDade Elementary Classical School', 99.0),
 ('James G Blaine Elementary School', 99.0),
 ('LaSalle Elementary Language Academy', 99.0),
 ('Mary E Courtenay Elementary Language Arts Center', 99.0),
 ('Northside College Preparatory High School', 99.0),
 ('Northside Learning Center High School', 99.0),
 ('Norwood Park Elementary School', 99.0),
 ('Oriole Park Elementary School', 99.0),
 ('Sauganash Elementary School', 99.0),
 ('Stephen Decatur Classical Elementary School', 99.0),
 ('Talman Elementary School', 99.0),
 ('Wildwood Elementary School', 99.0)]

Double-click **here** for the solution.

<!-- Solution:
In the previous problem we found out that the highest Safety Score is 99, so we can use that as an input in the where clause:

%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where Safety_Score = 99

or, a better way:

%sql select Name_of_School, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where \
  Safety_Score= (select MAX(Safety_Score) from CHICAGO_PUBLIC_SCHOOLS_DATA)


Correct answer: several schools with with Safety Score of 99.
-->


### Problem 4

##### What are the top 10 schools with the highest "Average Student Attendance"?


In [25]:
%sql select Name_of_School, AVERAGE_STUDENT_ATTENDANCE from CHICAGO_PUBLIC_SCHOOLS_DATA order by AVERAGE_STUDENT_ATTENDANCE desc limit 10;

 * sqlite:///RealWorldData.db
Done.


[('John Charles Haines Elementary School', '98.40%'),
 ('James Ward Elementary School', '97.80%'),
 ('Edgar Allan Poe Elementary Classical School', '97.60%'),
 ('Orozco Fine Arts & Sciences Elementary School', '97.60%'),
 ('Rachel Carson Elementary School', '97.60%'),
 ('Annie Keller Elementary Gifted Magnet School', '97.50%'),
 ('Andrew Jackson Elementary Language Academy', '97.40%'),
 ('Lenart Elementary Regional Gifted Center', '97.40%'),
 ('Disney II Magnet School', '97.30%'),
 ('John H Vanderpoel Elementary Magnet School', '97.20%')]

Double-click **here** for the solution.

<!-- Solution:

%sql select Name_of_School, Average_Student_Attendance from CHICAGO_PUBLIC_SCHOOLS_DATA \
    order by Average_Student_Attendance desc nulls last limit 10 

-->


### Problem 5

##### Retrieve the list of 5 Schools with the lowest Average Student Attendance sorted in ascending order based on attendance


In [26]:
%sql select Name_of_School, AVERAGE_STUDENT_ATTENDANCE from CHICAGO_PUBLIC_SCHOOLS_DATA order by AVERAGE_STUDENT_ATTENDANCE limit 5;

 * sqlite:///RealWorldData.db
Done.


[('Velma F Thomas Early Childhood Center', None),
 ('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Dyett High School', '62.50%'),
 ('Wendell Phillips Academy High School', '63.00%')]

Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 6

##### Now remove the '%' sign from the above result set for Average Student Attendance column


In [27]:
%sql select Name_of_School, replace(AVERAGE_STUDENT_ATTENDANCE, '%', '') from CHICAGO_PUBLIC_SCHOOLS_DATA order by AVERAGE_STUDENT_ATTENDANCE limit 5;

 * sqlite:///RealWorldData.db
Done.


[('Velma F Thomas Early Childhood Center', None),
 ('Richard T Crane Technical Preparatory High School', '57.90'),
 ('Barbara Vick Early Childhood & Family Center', '60.90'),
 ('Dyett High School', '62.50'),
 ('Wendell Phillips Academy High School', '63.00')]

Double-click **here** for a hint

<!--
Use the REPLACE() function to replace '%' with ''

-->


Double-click **here** for the solution.

<!-- Hint:

%sql SELECT Name_of_School, REPLACE(Average_Student_Attendance, '%', '') \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     order by Average_Student_Attendance \
     LIMIT 5

-->


### Problem 7

##### Which Schools have Average Student Attendance lower than 70%?


In [29]:
%sql select Name_of_School, AVERAGE_STUDENT_ATTENDANCE from CHICAGO_PUBLIC_SCHOOLS_DATA where AVERAGE_STUDENT_ATTENDANCE < 70 order by AVERAGE_STUDENT_ATTENDANCE;

 * sqlite:///RealWorldData.db
Done.


[('Richard T Crane Technical Preparatory High School', '57.90%'),
 ('Barbara Vick Early Childhood & Family Center', '60.90%'),
 ('Dyett High School', '62.50%'),
 ('Wendell Phillips Academy High School', '63.00%'),
 ('Orr Academy High School', '66.30%'),
 ('Manley Career Academy High School', '66.80%'),
 ('Chicago Vocational Career Academy High School', '68.80%'),
 ('Roberto Clemente Community Academy High School', '69.60%')]

Double-click **here** for a hint

<!--
The datatype of the "Average_Student_Attendance" column is varchar.
So you cannot use it as is in the where clause for a numeric comparison.
First use the CAST() function to cast it as a DECIMAL or DOUBLE
e.g. CAST("Column_Name" as DOUBLE)

-->


Double-click **here** for another hint

<!--
Don't forget the '%' age sign needs to be removed before casting
-->


Double-click **here** for the solution.

<!-- Solution:

%sql SELECT Name_of_School, Average_Student_Attendance  \
     from CHICAGO_PUBLIC_SCHOOLS_DATA \
     where CAST ( REPLACE(Average_Student_Attendance, '%', '') AS DOUBLE ) < 70 \
     order by Average_Student_Attendance
     

-->


### Problem 8

##### Get the total College Enrollment for each Community Area


In [32]:
%sql select sum(COLLEGE_ENROLLMENT) as TOTAL_ENROLLMENT, COMMUNITY_AREA_NAME from CHICAGO_PUBLIC_SCHOOLS_DATA group by COMMUNITY_AREA_NAME;

 * sqlite:///RealWorldData.db
Done.


[(6864, 'ALBANY PARK'),
 (4823, 'ARCHER HEIGHTS'),
 (1458, 'ARMOUR SQUARE'),
 (6483, 'ASHBURN'),
 (4175, 'AUBURN GRESHAM'),
 (10933, 'AUSTIN'),
 (1522, 'AVALON PARK'),
 (3640, 'AVONDALE'),
 (14386, 'BELMONT CRAGIN'),
 (1636, 'BEVERLY'),
 (3167, 'BRIDGEPORT'),
 (9647, 'BRIGHTON PARK'),
 (549, 'BURNSIDE'),
 (1568, 'CALUMET HEIGHTS'),
 (5042, 'CHATHAM'),
 (7086, 'CHICAGO LAWN'),
 (2085, 'CLEARING'),
 (4670, 'DOUGLAS'),
 (4568, 'DUNNING'),
 (5337, 'EAST GARFIELD PARK'),
 (5305, 'EAST SIDE'),
 (4600, 'EDGEWATER'),
 (910, 'EDISON PARK'),
 (6832, 'ENGLEWOOD'),
 (1431, 'FOREST GLEN'),
 (531, 'FULLER PARK'),
 (9915, 'GAGE PARK'),
 (4552, 'GARFIELD RIDGE'),
 (2809, 'GRAND BOULEVARD'),
 (4051, 'GREATER GRAND CROSSING'),
 (963, 'HEGEWISCH'),
 (3975, 'HERMOSA'),
 (8620, 'HUMBOLDT PARK'),
 (1930, 'HYDE PARK'),
 (7764, 'IRVING PARK'),
 (1755, 'JEFFERSON PARK'),
 (4287, 'KENWOOD'),
 (7055, 'LAKE VIEW'),
 (5615, 'LINCOLN PARK'),
 (4132, 'LINCOLN SQUARE'),
 (7351, 'LOGAN SQUARE'),
 (871, 'LOOP'),
 (7257, 'LOWER WEST SIDE'),
 (1552, 'MCKINLEY PARK'),
 (1317, 'MONTCLARE'),
 (3271, 'MORGAN PARK'),
 (2091, 'MOUNT GREENWOOD'),
 (3362, 'NEAR NORTH SIDE'),
 (1378, 'NEAR SOUTH SIDE'),
 (7975, 'NEAR WEST SIDE'),
 (7922, 'NEW CITY'),
 (7541, 'NORTH CENTER'),
 (5146, 'NORTH LAWNDALE'),
 (4210, 'NORTH PARK'),
 (6469, 'NORWOOD PARK'),
 (140, 'OAKLAND'),
 (786, 'OHARE'),
 (6954, 'PORTAGE PARK'),
 (1620, 'PULLMAN'),
 (1547, 'RIVERDALE'),
 (4068, 'ROGERS PARK'),
 (7020, 'ROSELAND'),
 (4043, 'SOUTH CHICAGO'),
 (1859, 'SOUTH DEERING'),
 (14793, 'SOUTH LAWNDALE'),
 (4543, 'SOUTH SHORE'),
 (4388, 'UPTOWN'),
 (4006, 'WASHINGTON HEIGHTS'),
 (2648, 'WASHINGTON PARK'),
 (3700, 'WEST ELSDON'),
 (5946, 'WEST ENGLEWOOD'),
 (2622, 'WEST GARFIELD PARK'),
 (4207, 'WEST LAWN'),
 (3240, 'WEST PULLMAN'),
 (8197, 'WEST RIDGE'),
 (9429, 'WEST TOWN'),
 (4206, 'WOODLAWN')]

Double-click **here** for a hint

<!--
Verify the exact name of the Enrollment column in the database
Use the SUM() function to add up the Enrollments for each Community Area
-->


Double-click **here** for another hint

<!--
Don't forget to group by the Community Area
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name 

-->


### Problem 9

##### Get the 5 Community Areas with the least total College Enrollment  sorted in ascending order


In [34]:
%sql select COMMUNITY_AREA_NAME, sum(COLLEGE_ENROLLMENT) as TOTAL_ENROLLMENT from CHICAGO_PUBLIC_SCHOOLS_DATA group by COMMUNITY_AREA_NAME order by sum(COLLEGE_ENROLLMENT) limit 5;

 * sqlite:///RealWorldData.db
Done.


[('OAKLAND', 140),
 ('FULLER PARK', 531),
 ('BURNSIDE', 549),
 ('OHARE', 786),
 ('LOOP', 871)]

Double-click **here** for a hint

<!--
Order the previous query and limit the number of rows you fetch
-->


Double-click **here** for the solution.

<!-- Solution:

%sql select Community_Area_Name, sum(College_Enrollment) AS TOTAL_ENROLLMENT \
   from CHICAGO_PUBLIC_SCHOOLS_DATA \
   group by Community_Area_Name \
   order by TOTAL_ENROLLMENT asc \
   LIMIT 5 

-->


### Problem 10

##### List 5 schools with lowest safety score.


In [38]:
%sql select NAME_OF_SCHOOL, Safety_Score from CHICAGO_PUBLIC_SCHOOLS_DATA where Safety_Score !='None' order by Safety_Score limit 5; 

 * sqlite:///RealWorldData.db
Done.


[('Edmond Burke Elementary School', 1.0),
 ("Luke O'Toole Elementary School", 5.0),
 ('George W Tilton Elementary School', 6.0),
 ('Foster Park Elementary School', 11.0),
 ('Emil G Hirsch Metropolitan High School', 13.0)]

Double-click **here** for the solution.

<!-- Solution SQLite:

%sql SELECT name_of_school, safety_score \
FROM CHICAGO_PUBLIC_SCHOOLS_DATA  where safety_score !='None' \
ORDER BY safety_score \
LIMIT 5

-->


### Problem 11

##### Get the hardship index for the community area which has College Enrollment of 4368


In [40]:
%sql select hardship_index from CENSUS_DATA CD, CHICAGO_PUBLIC_SCHOOLS_DATA CPS where CD.community_area_number = CPS.community_area_number and College_Enrollment = 4368;

 * sqlite:///RealWorldData.db
Done.


[(6.0,)]

Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%%sql 
select hardship_index from CENSUS_DATA CD, CHICAGO_PUBLIC_SCHOOLS_DATA CPS 
where CD.community_area_number = CPS.community_area_number 
and college_enrollment = 4368


-->


### Problem 12

##### Get the hardship index for the community area which has the highest value for College Enrollment


In [42]:
%sql select community_area_number, community_area_name, hardship_index from CENSUS_DATA \
   where community_area_number in \
   ( select community_area_number from CHICAGO_PUBLIC_SCHOOLS_DATA order by college_enrollment desc limit 1 )


 * sqlite:///RealWorldData.db
Done.


[(5.0, 'North Center', 6.0)]

Double-click **here** for the solution.

<!-- Solution:
NOTE: For this solution to work the CHICAGO_SOCIOECONOMIC_DATA table 
      as created in the last lab of Week 3 should already exist

%sql select community_area_number, community_area_name, hardship_index from CENSUS_DATA \
   where community_area_number in \
   ( select community_area_number from CHICAGO_PUBLIC_SCHOOLS_DATA order by college_enrollment desc limit 1 )
-->


## Summary

##### In this lab you learned how to work with a real word dataset using SQL and Python. You learned how to query columns with spaces or special characters in their names and with mixed case names. You also used built in database functions and practiced how to sort, limit, and order result sets, as well as used sub-queries and worked with multiple tables.


## Author

<a href="https://www.linkedin.com/in/ravahuja/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01" target="_blank">Rav Ahuja</a>

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                        |
| ----------------- | ------- | ----------------- | ----------------------------------------- |
| 2022-03-04        | 2.2     | Lakshmi Holla     | Made changes in markdown cells            |
| 2020-11-27        | 2.1     | Sannareddy Ramesh | Modified data sets and added new problems |
| 2020-08-28        | 2.0     | Lavanya           | Moved lab to course repo in GitLab        |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
